### PR TITLE
lp:1616098: ssh,scp,dh pick reachable address among all

### PR DESCRIPTION
**Pull request by dimitern**
Uses the new SSHClient API facade v2 (from #6468), if
available, and network.ReachableHostPort() (#6454) in
code shared between the ssh, scp, and debug-hooks CLI
commands, to select a reachable address from all known
addresses of the target (machine/unit/host).

Fixes http://pad.lv/1616098.

QA steps (live tested on both MAAS 1.9 with multiple-
NIC nodes and on lxd - see below):

  1. Using the tip of `develop`, bootstrap a lxd xenial controller.
  2. juju switch controller
  3. juju status 0    # note of the IP address (IP1)
  4. lxc exec \<juju-instance-id\> -- ip addr add \<IP2\>/24 dev eth0
(add an extra IP on machine-0, from the same subnet)
  5. lxc exec \<juju-instance-id\> -- systemctl restart jujud-machine-0.service
(bounce the controller agent to pick up the new IP)
  6. juju ssh 0             # should work
  7. sudo ip route add blackhole \<IP1\>/32
(make the preferred public IP unreachable from your laptop)
  8. juju --debug ssh 0     # should fail with 'Invalid argument'
  9. git checkout dimitern/lp-1616098-ssh+scp-alladdresses
(switch to this PR's branch and rebuild juju & jujud)
  10. juju ssh 0                    # should work now, using IP2
  11. juju scp 0:/etc/issue .  # as above
  12. juju deploy ubuntu --to 0
  13. juju debug-hooks ubuntu/0 # should work as well
(small delay ~10s observed on lxd, because of 'sudo')

### CODE REVIEW

**Comment by dimitern** [comment]
@nskaggs Why are my PR not picked up by the pre-check job (almost every one I had to poke to get a run)?

**Comment by dimitern** [comment]
!!build!!

**Comment by nskaggs** [comment]
@dimitern I'm not sure. Ping me on IRC and I can check the logs when you open a new one to see what the reason is.

File: cmd/juju/commands/ssh_common_test.go [hunk 1]
```diff
@@ -154,6 +220,7 @@ func (s *SSHCommonSuite) setupModel(c *gc.C) {
 	m := s.getMachineForUnit(c, u)
 	s.setAddresses(c, m)
 	s.setKeys(c, m)
+	s.setLinkLayerDevicesAddresses(c, m)
```
Comment (frobware) [inline cmd/juju/commands/ssh_common_test.go]: This stuck out - why does the change require this?

File: cmd/juju/commands/ssh_unix_test.go [hunk 2]
```diff
@@ -40,8 +46,21 @@ var sshTests = []struct {
 		},
 	},
 	{
-		about: "connect to machine 0 and pass extra arguments",
-		args:  []string{"0", "uname", "-a"},
+		about:      "connect to machine 0 (api v2)",
+		args:       []string{"0"},
+		dialWith:   dialerFuncFor("0.private", "0.public", "0.1.2.3"), // last one set on machine 0 eth0
```
Comment (frobware) [inline cmd/juju/commands/ssh_unix_test.go]: It's not clear what the comment means.

**Comment by frobware** [comment]
I did actually QA this change with the LXD provider and it worked as advertised. Subjectively, it also "feels" quicker when running `juju ssh 0`, for example.

**Comment by macgreagoir** [comment]
I can't bless it, but LGTM.

File: cmd/juju/commands/ssh_unix_test.go (hunk 2 above)
Comment (dimitern) [inline cmd/juju/commands/ssh_unix_test.go]: It tries to clarify where that "0.1.2.3" came from - it was set by suite.setLinkLayerDevicesAddresses() for suite.m (machine-0). I'll rephrase it to make it clearer.

**Comment by dimitern** [comment]
$$fixes-1616098$$

**Comment by dimitern** [comment]
$$retry-transient-failure$$

**Comment by dimitern** [comment]
$$windows-this-time$$

**Comment by dimitern** [comment]
$$again$$
